Units of Measure, Decimals and Date formatting

PeteBa opened the pull request:
Hi (and thanks for the great plugin).

I have submitted a pull request to add the following features into the status_panel plugin:

1. provide units of measure decorations (i.e. m/s, %, hours:minutes) to number metrics making them more descriptive. This is similar to the single-stat feature
2. control the number of decimal places for numbers. Either default to displaying all available decimals or lock to a specific precision
3. add support for Date/Time thresholds allowing metrics that hold milliseconds since Unix Epoch (UTC) to be presented in a human-friendly format (e.g. "YY-MM-DD mm:ss").

In order to achieve this, I have made a number of modifications to you code as follows:
- **Editor.html:** Modified the Editor to differentiate three Threshold types (Number, String and Date) each with a number of type specific form inputs to allow decorating the metric in the display panel.
- **Editor.html:** Number metrics can be decorated with the standard Grafana units of measure selection field and the number of Decimals set to "auto" for full precision or to a set number of decimal places.
- **Editor.html:** Date metrics will display in local timezone and can have warn/crit thresholds entered via the standard date picker  input form.
- **Editor.html:** The Warn/Crit input fields have modified to limit data entry to the specific type of Threshold i.e. String is type=text, Number is type=number and Date is type datetime-local.
- **Status_ctrl.js:** Imported the moment and kbn libraries to provide the support for Date/Time and Units-of-Measure transformations/calculations
- **Status_ctrl.js:** Extended the original Threshold valueHandler to support String, Number and Date Thresholds. Retains the same code logic by converting Dates to numbers before threshold comparisons.
- **Status.ctrl.js:** Leverage series.display_value to hold a fully formatted metric with decorations (i.e. "3.456 m/s" or "4 April '17") once threshold checks have been carried out. 
- **Status.ctrl.js:** Extended the upgradeOldVersions to map any existing Threshold series to one the new type specific versions (Number, String, Date) to avoid duplication choices.

I recognise that the changes above are more than cosmetic and would appreciate any feedback as happy to modify the PR to fit in with your design plans. I know that the last point to split out the Thresholds into three types and depreciate the original may warrant some discussion.

PeteBa commented:
I have rebased this to reflect the latest master commits and fixed a bug so that Date metrics are displayed correctly when switching Grafana settings between browser to utc timezones.

TomerAmirV commented on src/status_ctrl.js:
```diff
@@ -273,6 +301,22 @@ export class StatusPluginCtrl extends MetricsPanelCtrl {
+	}
+
+	formatDisplayValue(value, target) {
+        if (target.valueHandler == "Number Threshold" && StatusPluginCtrl.isFloat(value)) {
```
The `isFloat` call causes exceptions when a `null` value is passed to it. please fix the method, or make sure that you do not pass null

PeteBa commented:
Thanks for the feedback on the pull request. I have rebased the PR to the latest master and addressed the issue you highlighted.

The code should now deal with the edge conditions where under certain circumstances (e.g. a query returning strings that then gets aggregated using Avg/Max/Min) the measurement.display_value can be NaN or Null. This is now trapped and handled within the formatDisplayValue() logic.

I am happy to change and improve the code if you have any feedback or test cases that I should run.

PeteBa commented:
Hi Guys, just just wanted to check in on PR #55 and #58. Are these things you are interested in or perhaps they don't fit with what you want from the plug-in? Totally understand either way. Cheers, Pete

PeteBa commented:
Hi, I have responded to the changes requested some time ago now. Could you please either provide further feedback or close the PR if not appropriate to merge. Many thanks. Peter

TomerAmirV commented:
Hi @PeteBa,
Sorry it took so long, had other tasks so I did not have time to check this plugin.
The changes look good now, but I have one last small request, please make sure that the default decimals value will be 2 or 3 for backwards compatibility, and after that I'll merge your pull request

Awesome job!

PeteBa commented:
Hey thanks for responding on this. I have rebased onto current master and changed decimal default to two. Cheers, Pete

PeteBa commented:
Thanks for merging this. Given that this modifies the panel json configuration (to support the new threshold types) I think it would be a good idea to include a warning to save/export before upgrading. Thanks again.

TomerAmirV commented:
Sounds like a good idea... maybe it would be best to add a general warning for upgrading

TomerAmirV commented:
https://github.com/grafana/grafana-plugin-repository/pull/158
